### PR TITLE
[idea] #334 Multiple results with same file name leads to collisions.

### LINKED ISSUE #329: NPE while using the plugin 2
null
java.lang.NullPointerException
	at com.imaginea.kodebeagle.util.WindowEditorOps$2.run(WindowEditorOps.java:51)
	at com.intellij.openapi.application.RunResult.run(RunResult.java:38)
	at com.intellij.openapi.command.WriteCommandAction$2$1.run(WriteCommandAction.java:114)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:931)
	at com.intellij.openapi.command.WriteCommandAction$2.run(WriteCommandAction.java:111)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:124)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:99)
	at com.intellij.openapi.command.WriteCommandAction.performWriteCommandAction(WriteCommandAction.java:108)
	at com.intellij.openapi.command.WriteCommandAction.execute(WriteCommandAction.java:80)
	at com.imaginea.kodebeagle.util.WindowEditorOps.setWriteStatus(WindowEditorOps.java:48)
	at com.imaginea.kodebeagle.util.EditorDocOps.getVirtualFile(EditorDocOps.java:274)
	at com.imaginea.kodebeagle.action.RefreshAction$CodePaneTinyEditorExpandLabelMouseListener.mouseClicked(RefreshAction.java:485)
	at java.awt.Component.processMouseEvent(Component.java:6519)
	at javax.swing.JComponent.processMouseEvent(JComponent.java:3312)
	at java.awt.Component.processEvent(Component.java:6281)
	at java.awt.Container.processEvent(Container.java:2229)
	at java.awt.Component.dispatchEventImpl(Component.java:4872)
	at java.awt.Container.dispatchEventImpl(Container.java:2287)
	at java.awt.Component.dispatchEvent(Component.java:4698)
	at java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4832)
	at java.awt.LightweightDispatcher.processMouseEvent(Container.java:4501)
	at java.awt.LightweightDispatcher.dispatchEvent(Container.java:4422)
	at java.awt.Container.dispatchEventImpl(Container.java:2273)
	at java.awt.Window.dispatchEventImpl(Window.java:2719)
	at java.awt.Component.dispatchEvent(Component.java:4698)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:747)
	at java.awt.EventQueue.access$300(EventQueue.java:103)
	at java.awt.EventQueue$3.run(EventQueue.java:706)
	at java.awt.EventQueue$3.run(EventQueue.java:704)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:76)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:87)
	at java.awt.EventQueue$4.run(EventQueue.java:720)
	at java.awt.EventQueue$4.run(EventQueue.java:718)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:76)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:717)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:734)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:565)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:382)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:242)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:161)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:150)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:146)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:138)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:91)

### LINKED ISSUE #334: If two different project have same filename in result sets for a search wrong/junk results are displayed.
screenshot from 2015-08-05 09 51 16

### PR DESCRIPTION
**Pull request by ScrapCodes**
1) Generate sha of project name to solve both long file name issue in windows.
2) Solve the problem of name collision by using full filename.

It also seems to solve the problem of NPE while using the plugin,
VFSDirectoryImpl will not adopt a file if it is accessed and later deleted.

So removing deleteOnExist seems to fix the problem.

This has a negative side that it may lead to using some extra space in a users temp
directory. But that should be hopefully okay.

closes #334
closes #329

### CODE REVIEW

**Comment by ScrapCodes** [comment]
@amit-shar and @prudhvije Please take a look.

File: plugins/idea/kodebeagleidea/checkstyle-config.xml [hunk 1]
```diff
@@ -174,7 +174,9 @@
         <module name="HiddenField"/>
         <module name="IllegalInstantiation"/>
         <module name="InnerAssignment"/>
-        <module name="MagicNumber"/>
+        <module name="MagicNumber">
+         <property name="ignoreNumbers" value="-2, -1, 0, 1, 2, 3, 10"/>
```
Comment (prudhvije) [inline plugins/idea/kodebeagleidea/checkstyle-config.xml]: Why 10 is excluded? -2, -1, 0 are by default not magic numbers right?

File: plugins/idea/kodebeagleidea/src/main/java/com/imaginea/kodebeagle/util/EditorDocOps.java [hunk 2]
```diff
@@ -200,35 +203,41 @@
...
-        file.deleteOnExit();
-        VirtualFile virtualFile = LocalFileSystem.getInstance().findFileByPath(filePath);
-        windowEditorOps.setWriteStatus(virtualFile, false);
         return virtualFile;
     }
 
+    @NotNull
+    private String getDigestAsString(final String trimmedFileName)
+            throws UnsupportedEncodingException, NoSuchAlgorithmException {
+        MessageDigest crypt = MessageDigest.getInstance("SHA-1");
```
Comment (prudhvije) [inline plugins/idea/kodebeagleidea/src/main/java/com/imaginea/kodebeagle/util/EditorDocOps.java]: You should use a constant here.

File: plugins/idea/kodebeagleidea/src/main/java/com/imaginea/kodebeagle/util/EditorDocOps.java [hunk 3]
```diff
@@ -200,35 +203,41 @@
...
+        crypt.reset();
+        crypt.update(trimmedFileName.getBytes("utf8"));
```
Comment (prudhvije) [inline plugins/idea/kodebeagleidea/src/main/java/com/imaginea/kodebeagle/util/EditorDocOps.java]: And even here.

File: plugins/idea/kodebeagleidea/src/main/java/com/imaginea/kodebeagle/action/RefreshAction.java [hunk 4]
```diff
@@ -493,17 +494,24 @@ public CodePaneTinyEditorExpandLabelMouseListener(final String pDisplayFileName,
...
+                Document document =
+                        EditorFactory.getInstance().createDocument(windowObjects.
+                                getFileNameContentsMap().get(fileName));
+                editorDocOps.addHighlighting(windowObjects.
+                        getFileNameNumbersMap().get(fileName), document);
+                editorDocOps.gotoLine(windowObjects.
+                        getFileNameNumbersMap().get(fileName).get(0), document);
+            } catch (Exception exception) {
+                exception.printStackTrace();
+            }
```
Comment (prudhvije) [inline plugins/idea/kodebeagleidea/src/main/java/com/imaginea/kodebeagle/action/RefreshAction.java]: Notify user something about this? He tries to click on that filename and does nothing in response? whole UX is destroyed here.

File: plugins/idea/kodebeagleidea/src/main/java/com/imaginea/kodebeagle/util/EditorDocOps.java [hunk 5]
```diff
@@ -200,35 +203,41 @@
...
+                                            final String contents)
+            throws IOException, NoSuchAlgorithmException {
+
+        final String tempDir = System.getProperty(JAVA_IO_TMP_DIR);
+        final String trimmedFileName =
+                FileUtil.sanitizeFileName(StringUtil.trimEnd(fileName, displayFileName));
+        final String digest = getDigestAsString(trimmedFileName);
+        final String filePath = tempDir + File.separatorChar + digest
+                + File.separatorChar + displayFileName;
+        final File file = new File(filePath);
```
Comment (prudhvije) [inline plugins/idea/kodebeagleidea/src/main/java/com/imaginea/kodebeagle/util/EditorDocOps.java]: isn't it stringbuilder better? too may strings to add.

**Comment by prudhvije** [comment]
Getting this: 
java.lang.NullPointerException
	at com.imaginea.kodebeagle.util.WindowEditorOps$2.run(WindowEditorOps.java:51)
	at com.intellij.openapi.application.RunResult.run(RunResult.java:38)
	at com.intellij.openapi.command.WriteCommandAction$2$1.run(WriteCommandAction.java:114)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:931)
	at com.intellij.openapi.command.WriteCommandAction$2.run(WriteCommandAction.java:111)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:124)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:99)
	at com.intellij.openapi.command.WriteCommandAction.performWriteCommandAction(WriteCommandAction.java:108)
	at com.intellij.openapi.command.WriteCommandAction.execute(WriteCommandAction.java:80)
	at com.imaginea.kodebeagle.util.WindowEditorOps.setWriteStatus(WindowEditorOps.java:48)
	at com.imaginea.kodebeagle.util.EditorDocOps.getVirtualFile(EditorDocOps.java:226)
	at com.imaginea.kodebeagle.action.RefreshAction$CodePaneTinyEditorExpandLabelMouseListener.mouseClicked(RefreshAction.java:500)
	at java.awt.Component.processMouseEvent(Component.java:6519)
	at javax.swing.JComponent.processMouseEvent(JComponent.java:3321)
	at java.awt.Component.processEvent(Component.java:6281)
	at java.awt.Container.processEvent(Container.java:2229)
	at java.awt.Component.dispatchEventImpl(Component.java:4872)
	at java.awt.Container.dispatchEventImpl(Container.java:2287)
	at java.awt.Component.dispatchEvent(Component.java:4698)
	at java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4832)
	at java.awt.LightweightDispatcher.processMouseEvent(Container.java:4501)
	at java.awt.LightweightDispatcher.dispatchEvent(Container.java:4422)
	at java.awt.Container.dispatchEventImpl(Container.java:2273)
	at java.awt.Window.dispatchEventImpl(Window.java:2719)
	at java.awt.Component.dispatchEvent(Component.java:4698)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:747)
	at java.awt.EventQueue.access$300(EventQueue.java:103)
	at java.awt.EventQueue$3.run(EventQueue.java:706)
	at java.awt.EventQueue$3.run(EventQueue.java:704)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:76)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:87)
	at java.awt.EventQueue$4.run(EventQueue.java:720)
	at java.awt.EventQueue$4.run(EventQueue.java:718)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:76)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:717)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:734)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:565)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:382)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:242)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:161)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:150)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:146)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:138)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:91)

Havent' touched any temp files.

**Comment by ScrapCodes** [comment]
@prudhvije and @amit-shar  I have tried to fix some of the issues, Please take another look !

**Comment by ScrapCodes** [comment]
@prudhvije Do you have any more comments ?

File: plugins/idea/kodebeagleidea/checkstyle-config.xml (hunk 1 above)
Comment (ScrapCodes) [inline plugins/idea/kodebeagleidea/checkstyle-config.xml]: I am passing 10 in string.sublist, it did not appear to me as magic number. so wanted ignore it.

**Pull request closed without merging**